[link] Support adding bank account using FC

lng-stripe opened the pull request:
# Summary

Adds to Link the ability to add a bank account as a payment method using the Financial Connections SDK. Supports both FC Web, FC Native, and FC Lite.

This PR is bigger than anticipated because multiple related fixes and refactorings were required. See PR comments for details.

# Motivation

https://jira.corp.stripe.com/browse/LINK_MOBILE-12

# Testing
- [x] Added tests
- [x] Modified tests
- [x] Manually verified

# Screenshots

https://github.com/user-attachments/assets/befed42d-223f-4cb8-ba19-c30806e07cb3

# Changelog
<!-- Is this a notable change that affects users? If so, add a line to `CHANGELOG.md` and prefix the line with one of the following:
    - [Added] for new features.
    - [Changed] for changes in existing functionality.
    - [Deprecated] for soon-to-be removed features.
    - [Removed] for now removed features.
    - [Fixed] for any bug fixes.
    - [Security] in case of vulnerabilities.
-->

tillh-stripe commented on financial-connections/src/main/java/com/stripe/android/financialconnections/presentation/FinancialConnectionsSheetNativeViewModel.kt:
```diff
@@ -340,7 +341,7 @@ internal class FinancialConnectionsSheetNativeViewModel @Inject constructor(
 
                     session.hasAValidAccount() -> {
-                        if (state.isLinkWithStripe) {
+                        if (state.flowType == FinancialConnectionsSheetFlowType.ForInstantDebits) {
```
Yep, that looks right to me!

lng-stripe commented on payments-model/src/main/java/com/stripe/android/model/ConsumerPaymentDetails.kt:
```diff
@@ -19,7 +19,7 @@ data class ConsumerPaymentDetails(
         open val type: String,
         open val nickname: String?,
-    ) : Parcelable {
+    ) : StripeModel {
```
Hm, might be vestigial, I'll undo.

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/account/DefaultLinkAccountManager.kt:
```diff
@@ -273,6 +279,25 @@ internal class DefaultLinkAccountManager @Inject constructor(
+                bankAccountId = bankAccountId,
+                userEmail = linkAccount.email,
+                consumerSessionClientSecret = linkAccount.clientSecret,
+                consumerPublishableKey = linkAccount.consumerPublishableKey.takeIf { config.passthroughModeEnabled },
```
Removed -- only need merchant PK.

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/wallet/AddBankAccountState.kt:
```diff
@@ -0,0 +1,17 @@
+
+    data object PresentingFinancialConnections : AddBankAccountState
+
+    data object ProcessingFinancialConnectionsResult : AddBankAccountState
```
I consolidated most into `data class Processing(val configToPresent: FinancialConnectionsSheetConfiguration? = null)` to simplify

lng-stripe commented:
@tillh-stripe I addressed your comments (see comments/commits). In addition, I also:
 * Rebased master and fixed merge conflicts
 * Added a feature flag to force enable FC native for testing
 * Refactored `WalletScreen()` composable a bit and added tests for showing the add payment method menu.

carlosmuvi-stripe commented on payments-core/src/main/java/com/stripe/android/payments/financialconnections/FinancialConnectionsAvailability.kt:
```diff
@@ -1,9 +1,24 @@
     Lite
 }
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+fun FinancialConnectionsAvailability.getIntentBuilder(
+    context: Context
+): (FinancialConnectionsSheetActivityArgs) -> Intent {
+    return when (this) {
+        FinancialConnectionsAvailability.Full -> intentBuilder(context)
+        FinancialConnectionsAvailability.Lite -> liteIntentBuilder(context)
+    }
+}
```
Thanks!

carlosmuvi-stripe commented on payments-model/src/main/java/com/stripe/android/model/LinkAccountSession.kt:
```diff
@@ -0,0 +1,12 @@
+import kotlinx.parcelize.Parcelize
+
+@Parcelize
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+data class LinkAccountSession(
+    val id: String,
+    val clientSecret: String,
+) : StripeModel
```
You can probably use [this model](https://github.com/stripe/stripe-android/blob/master/payments-core/src/main/java/com/stripe/android/model/FinancialConnectionsSession.kt#L13-L12) that already exists.

carlosmuvi-stripe commented on payments-model/src/main/java/com/stripe/android/model/parsers/LinkAccountSessionJsonParser.kt:
```diff
@@ -0,0 +1,21 @@
+import com.stripe.android.model.LinkAccountSession
+import org.json.JSONObject
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+object LinkAccountSessionJsonParser :
+    ModelJsonParser<LinkAccountSession> {
+
+    private const val FIELD_ID = "id"
+    private const val FIELD_CLIENT_SECRET = "client_secret"
+
+    override fun parse(json: JSONObject): LinkAccountSession {
+        return LinkAccountSession(
+            id = json.getString(FIELD_ID),
+            clientSecret = json.getString(FIELD_CLIENT_SECRET),
+        )
+    }
+}
```
parser for the above exists as well : )

carlosmuvi-stripe commented on payments-model/src/main/java/com/stripe/android/repository/ConsumersApiService.kt:
```diff
@@ -434,6 +471,11 @@ class ConsumersApiServiceImpl(
          */
         private val createPaymentDetails: String = getApiUrl("consumers/payment_details")
 
+        /**
+         * @return `https://api.stripe.com/v1/consumers/link_account_sessions`
+         */
+        private val createLinkAccountSession: String = getApiUrl("consumers/link_account_sessions")
+
```
noticed that this exists in StripeApiRepository: https://github.com/stripe/stripe-android/blob/master/payments-core/src/main/java/com/stripe/android/networking/StripeApiRepository.kt#L1950-L1955

We should probably remove it from there (it's unused)

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/wallet/WalletScreen.kt:
```diff
@@ -732,6 +779,28 @@ private fun AlertMessage(
     )
 }
 
+// We can't use FC's `rememberFinancialConnectionsSheet` because it doesn't support FC Lite.
+@Composable
+private fun rememberFinancialConnectionsSheetInternal(
+    financialConnectionsAvailability: FinancialConnectionsAvailability?,
+    callback: (FinancialConnectionsSheetResult) -> Unit
+): FinancialConnectionsSheetForDataLauncher? {
+    financialConnectionsAvailability ?: return null
+
+    val context = LocalContext.current
+    val contract = remember(context, financialConnectionsAvailability) {
+        FinancialConnectionsSheetForDataContract(
+            intentBuilder = financialConnectionsAvailability.getIntentBuilder(context)
+        )
+    }
+    val activityResultLauncher = rememberLauncherForActivityResult(contract, callback)
+    return remember(activityResultLauncher) {
+        FinancialConnectionsSheetForDataLauncher(
+            activityResultLauncher
+        )
+    }
+}
```
good catch - when we make link standalone public we'll likely need to publish this

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/wallet/WalletScreen.kt:
```diff
@@ -160,7 +188,24 @@ internal fun WalletBody(
                     onLogoutClicked = onLogoutClicked,
                     onSetDefaultClicked = onSetDefaultClicked,
-                    onAddNewPaymentMethodClicked = onAddNewPaymentMethodClicked,
+                    onAddNewPaymentMethodClicked = {
+                        if (state.addPaymentMethodOptions.size == 1) {
+                            onAddPaymentMethodOptionClicked(state.addPaymentMethodOptions[0])
+                        } else {
```
guessing we don't want to show the modal if there's just one PM option, is that the behavior across platforms?

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/PaymentOptionsViewModel.kt:
```diff
@@ -237,7 +237,7 @@ internal class PaymentOptionsViewModel @Inject constructor(
             if (shouldShowLinkConfiguration) {
                 linkPaymentLauncher.present(
-                    configuration = linkState.configuration,
+                    configuration = linkState!!.configuration,
```
why is this needed?

carlosmuvi-stripe reviewed (commented):
looking great! just a few comments.
